dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs

```diff
@@ -0,0 +1,516 @@
+using Microsoft.SemanticKernel.Connectors.Memory.Weaviate.Http.ApiSchema;
+using Microsoft.SemanticKernel.Connectors.Memory.Weaviate.Model;
+using Microsoft.SemanticKernel.Memory;
+
+namespace Microsoft.SemanticKernel.Connectors.Memory.Weaviate;
+
+/// <summary>
+/// An implementation of <see cref="IMemoryStore" /> for Weaviate.
+/// </summary>
+/// <remarks>The Embedding data is saved to Weaviate instance specified in the constructor.
+/// The embedding data persists between subsequent instances and has similarity search capability.
+/// </remarks>
+// ReSharper disable once ClassWithVirtualMembersNeverInherited.Global
+public class WeaviateMemoryStore : IMemoryStore, IDisposable
+{
+    // Regex to ensure Weaviate class names confirm to the naming convention
+    // https://weaviate.io/developers/weaviate/configuration/schema-configuration#class
+    private static readonly Regex s_classNameRegEx = new("[^0-9a-zA-Z]+", RegexOptions.Compiled);
+
+    private static readonly JsonSerializerOptions s_jsonSerializerOptions = new()
+    {
+        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
+    };
+
+    private readonly HttpClient _httpClient;
+    private readonly bool _isSelfManagedHttpClient;
+
+    private readonly ILogger _log;
+    private bool _disposed;
+
+    /// <summary>
+    ///     Constructor for a memory store backed by Weaviate
+    /// </summary>
+    public WeaviateMemoryStore(string scheme, string host, int port, string? apiKey = null, HttpClient? httpClient = null, ILogger? logger = null)
+    {
+        Verify.ArgNotNullOrEmpty(scheme, "Scheme cannot be null or empty");
+        Verify.ArgNotNullOrEmpty(host, "Endpoint cannot be null or empty");
+
+        this._log = logger ?? NullLogger<WeaviateMemoryStore>.Instance;
+        if (httpClient == null)
+        {
+            if (string.IsNullOrEmpty(apiKey))
+            {
+                this._httpClient = new();
+            }
+            else
+            {
+                this._httpClient = new()
+                {
+                    DefaultRequestHeaders =
+                    {
+                        {
+                            "authorization", apiKey
+                        }
+                    }
+                };
+            }
+
+            // If not passed an HttpClient, then it is the responsibility of this class
+            // to ensure it is cleared up in the Dispose() method.
+            this._isSelfManagedHttpClient = true;
+        }
+        else
+        {
+            this._httpClient = httpClient;
+        }
+
+        this._httpClient.BaseAddress = new($"{scheme}://{host}:{port}/v1/");
+    }
+
+    public void Dispose()
+    {
+        this.Dispose(true);
+        GC.SuppressFinalize(this);
+    }
+
+    /// <inheritdoc />
+    public async Task CreateCollectionAsync(string collectionName, CancellationToken cancellationToken = default)
```

Comment:
Should we check in this method whether collection already exists or not?

Reply:
The `DoesCollectionExistAsync` exists on the interface which, to my mind, indicates that the checking as to whether a collection exists is the responsibility of the caller?

Otherwise this method should be named `CreateCollectionIfNotExistsAsync` - right?